DM-12635: Subpackage for converting Gen2->Gen3.

TallJimbo opened the pull request:
The conversion code here has two high-level drivers:
 - The `ConversionWalker` class (`walker.py`) extracts everything we'll need for the conversion from a Gen2 repo.  It's basically complete, and I've done some one-off testing it on `ci_hsc` (check out `ci_hsc`, setup and build, and start by calling `ConversionWalker.tryRoot` on `$CI_HSC_DIR/DATA`).
 - The `ConversionWriter` class (`writer.py`) adds entries to a `Registry` and `Datastore` - or it would if all of the actual calls to do so weren't just "TODO" comment placeholders

As for those TODO placeholders, there are a ton of them, and I think they cover most of what needs to be done to get things working.  Beyond that, we'll also eventually need a high-level driver that can be run easily from the command-line, but that doesn't necessarily need to be done right now.

TallJimbo committed:
Add subpackage for converting Gen2->Gen3.

TallJimbo committed:
Add PhysicalFilter to Exposure.

TallJimbo committed:
Fix Exposure/Visit timestamp relationships.

TallJimbo committed:
Minor fixes for schema yaml.

TallJimbo committed:
Add schema entries from VisitInfo.

afw.image.VisitInfo provides a good initial guess at
what metadata we'll want for Visits and Exposures.

TallJimbo committed:
Add seeing to Visit table.

timj commented:
Important note: datastore doesn't yet write the formatter information to somewhere permanent. It uses a shim at the moment that passes tests but forgets immediately the process ends. We need a more permanent fix in there.

TallJimbo commented:
Good to know.  I think we'd get a lot of the benefits from just getting this to write to `Registry`, as that'd let us easily put together realistically complex test datasets for it.

pschella commented on python/lsst/daf/butler/registries/sqlRegistry.py:
```diff
@@ -508,6 +510,7 @@ def addRun(self, run):
                                                         collection=run.collection,
                                                         environment_id=None,  # TODO add environment
                                                         pipeline_id=None))    # TODO add pipeline
+        # TODO: set given Run's 'id' attribute.
```
No, this is already done in the `Execution` part.

pschella commented on python/lsst/daf/butler/core/registry.py:
```diff
@@ -135,6 +135,7 @@ def addDataset(self, ref, uri, components, run, producer=None):
             If a Dataset with the given `DatasetRef` already exists in the
             given Collection.
         """
+        # TODO: this signature no longer agrees with SqlRegistry.addDataset.
```
One of many I'm afraid. I'm planning to copy them all over after things have settled a bit.

TallJimbo replied:
Yes, this TODO has already been removed.

pschella commented on python/lsst/daf/butler/gen2convert/extractor.py:
```diff
@@ -0,0 +1,123 @@
+
+from .structures import Gen2Dataset
+
+TEMPLATE_RE = re.compile(r'\%\((?P<name>\w+)\).*?(?P<type>[idrs])')
```
Would be nice to add an example comment for what this matches.

pschella commented on python/lsst/daf/butler/gen2convert/writer.py:
```diff
@@ -0,0 +1,174 @@
+        self.insertSkyMaps(registry)
+        self.insertObservations(registry)
+        self.insertDatasets(registry, datastore)
+        # TODO: associate parent repo datasets with child repo Collections
```
Is this `Registry.merge`? Or just `Registry.associate`?

TallJimbo committed:
Remove table for AbstractFilter.

We don't really have an entity responsible for creating
AbstractFilters, since they're not a associated with a
Camera or a SkyMap.  It's probably best to just let them
be "created" when first referenced, especially since we
don't actually need to store any metadata associated with
them.

TallJimbo committed:
Add butler_test_repository to .gitignore

TallJimbo committed:
Add interface for generic SQL queries to SqlRegistry.

TallJimbo committed:
Add __repr__ for Run.

TallJimbo committed:
Avoid nested connection contexts in addDataset.

Having addDataset call associate leads to "locked database"
errors, apparently because the transaction context managers
aren't nesting the way SQLAlchemy's documentation claims
they should.  This should be investigated further, but this
workaround lets us move forward for now.

TallJimbo committed:
Add API ingesting existing files into PosixDatastore.

TallJimbo commented:
This is now as far along as I think make sense on this ticket.  Next steps will be making `Datastore` records persistent and adding concrete `StorageClasses` and `Formatters`.

pschella commented on config/gen2convert.yaml:
```diff
@@ -0,0 +1,51 @@
+    # entries; options here say how to get those from a Gen2 repo.
+    VisitInfo:
+      # The Gen2 DatasetType to read when trying to create a VisitInfo.
+      # (we actually add a "_md" suffix, because we just read the metadata).
```
Something we don't (yet) have in Gen3. Unless you mean `findDataUnitEntry` type things here.

TallJimbo replied:
I was thinking we'd handle the metadata as another `Exposure` component dataset.

pschella commented on config/gen2convert.yaml:
```diff
@@ -0,0 +1,51 @@
+  # its Run; this will be the one corresponding to the Gen2 repository
+  # that originally contained the file, unless that has been overridden.
+  raw/HSC: 1
+  ref/ps1_pv3_3pi_20170110: 2
```
We may want to allow for a `None` default that will generate a new `Run`.

TallJimbo replied:
That default is already in place; if nothing here matches, we'll make a new `Run`, and use the repo root at its `Collection` name.

pschella commented on config/registry/default_schema.yaml:
```diff
@@ -195,14 +181,76 @@ dataUnits:
+          doc: >
+            Timestamp of the end of the Visit.  This should be the
+            same as the datetime_end of the last Exposure associated
+            with this Visit.
```
How is this enforced? Should it be?

TallJimbo replied:
The most important place to enforce it is probably the raw data ingest script.  Since it (and a number of other constraints) can probably be verified via SQL, it might make sense to have a script we can run explicitly to check these things.  I doubt trying to add complex table constraints is worthwhile; I would guess those are pretty RDBMS-specific.

pschella commented on config/registry/default_schema.yaml:
```diff
@@ -343,8 +401,23 @@ dataUnits:
+        -
+          name: rot_angle
+          type: float
+          doc: Rotation angle of the focal plane w.r.t. nominal
```
period

pschella commented on python/lsst/daf/butler/datastores/posixDatastore.py:
```diff
@@ -283,6 +283,23 @@ def put(self, inMemoryDataset, ref):
+        self.ingest(path, ref, formatter=formatter)
+
+    def ingest(self, path, ref, formatter=None):
+        """Record that a Datset with the given `DatasetRef` exists in the store.
```
Dataset

pschella commented on python/lsst/daf/butler/datastores/posixDatastore.py:
```diff
@@ -292,7 +309,7 @@ def put(self, inMemoryDataset, ref):
 
         # Associate this dataset with the formatter for later read.
-        fileInfo = StoredFileInfo(formatter, path, storageClass)
+        fileInfo = StoredFileInfo(formatter, path, ref.datasetType.storageClass)
```
How was this working before?

TallJimbo replied:
This code was moved out of `put`, which had a `storageClass` local variable pointing to the same object.

pschella commented on python/lsst/daf/butler/gen2convert/extractor.py:
```diff
@@ -0,0 +1,159 @@
+__all__ = ("Extractor",)
+
+
+TEMPLATE_RE = re.compile(r'\%\((?P<name>\w+)\).*?(?P<type>[idrs])')
```
Comment explaining what it matches?

pschella commented on python/lsst/daf/butler/gen2convert/extractor.py:
```diff
@@ -0,0 +1,159 @@
+        except StopIteration:
+            pass
+        terms.append(re.escape(template[m1.end():]))
+        return cls(datasetType=datasetType, regex=re.compile("".join(terms)))
```
Ok, this function definitely would benefits from some more comments. In particular describing the role of `m1` and `m2`. I also suspect that this would be clearer when setting `terms=None` for the first case and then using a single `for else` block for the looping of both cases.

pschella replied:
Also with parsing in general it is always convenient to have an example of the to-be-matched string in the documentation somewhere.

TallJimbo replied:
I had previously convinced myself that I couldn't use a for loop here because I needed to have both the current and next element available together, but of course that's silly.  Thanks for the extra push to simplify it; it's much better now.

pschella commented on config/gen2convert.yaml:
```diff
@@ -0,0 +1,51 @@
+      repl: "u/"
+    -
+      pattern: "^(.+)/rerun/"
+      repl: "shared/"
```
Are we inventing a namespacing convention for collections? Michelle might want input on this.

TallJimbo replied:
We're inventing a corner of a provisional one.  We can revisit later.

pschella commented on python/lsst/daf/butler/gen2convert/extractor.py:
```diff
@@ -0,0 +1,159 @@
+    Parameters
+    ----------
+    repo : `Gen2Repo`
+        Structure describing the repository this Extractor will process.
```
`Attributes` too?

TallJimbo replied:
Does that actually work?  I thought we had to do that nonsense with the class attributes that get overridden by instance attributes (and frankly I don't think that's worth it).

pschella commented on python/lsst/daf/butler/gen2convert/extractor.py:
```diff
@@ -0,0 +1,159 @@
+
+
+class Extractor:
+    """An object that parses Gen2 paths into Gen2Dataset instance for a
```
instance? instances?

pschella replied:
callable object?

pschella commented on python/lsst/daf/butler/gen2convert/translators.py:
```diff
@@ -0,0 +1,275 @@
+        handler.
+    gen3unit : `str`
+        Name of the Gen3 DataUnit associated with `gen3key`.
+    """
```
An example would be nice. (e.g. `camera`, `Camera`).

pschella commented on python/lsst/daf/butler/gen2convert/translators.py:
```diff
@@ -0,0 +1,275 @@
+        SkyMap instance used to define any Tract or Patch DataUnits.
+    skyMapName : `str`
+        Gen3 SkyMap DataUnit name to be associated with any Tract or Patch
+        DataUnits.
```
Why is this not `skyMap.name`?

TallJimbo replied:
Maybe it should be, in the future, but `BaseSkyMap` (and its subclasses) are not new, and I'm a little wary of adding a `name` attribute since it can't be derived from the SkyMap's actual contents.

pschella commented on python/lsst/daf/butler/gen2convert/translators.py:
```diff
@@ -0,0 +1,275 @@
+    # Rules used to match Handlers when constring a Translator.
+    # outer key is camera name, or None for any
+    # inner key is DatasetType name, or None for any
+    # values are 2-tuples of (frozenset(gen2keys), handler, consume)
```
How is this a 2-tuple?

TallJimbo replied:
Historically, before I added `consume` :-)

pschella commented on python/lsst/daf/butler/gen2convert/translators.py:
```diff
@@ -0,0 +1,275 @@
+
+    @classmethod
+    def addRule(cls, handler, camera=None, datasetTypeName=None, gen2keys=(), consume=True):
+        """Add a Handler and an associated matching rule.
```
`KeyHandler`

pschella commented on python/lsst/daf/butler/gen2convert/translators.py:
```diff
@@ -0,0 +1,275 @@
+        return frozenset(h.gen3unit for h in self.handlers)
+
+
+class ConstantKeyHandler(KeyHandler):
```
You may want to group these with the `KeyHandler`? Without the `Translator` in between.

pschella commented on python/lsst/daf/butler/gen2convert/translators.py:
```diff
@@ -0,0 +1,275 @@
+
+
+class PatchKeyHandler(KeyHandler):
+    """A KeyHandler for Patchs."""
```
Patches?

pschella commented on python/lsst/daf/butler/gen2convert/walker.py:
```diff
@@ -0,0 +1,329 @@
+    necessary to create a Gen3 'snapshot' view to them.
+
+    ConversionWalkers attempt to only include state that does not involve any user
+    choices about to perform the conversion; anything that does is handled
```
about how to

pschella commented on python/lsst/daf/butler/gen2convert/walker.py:
```diff
@@ -0,0 +1,329 @@
+
+        If the path appears to be a Gen2 Data Repository, a Gen2Data instance
+        representing it will be added to `self.found` (unless it is already there)
+        and returned.  `tryRoot()` also be called recursively on all parent
```
can be? will be? called recursively...

TallJimbo replied:
`will be` (fixed)

pschella commented on python/lsst/daf/butler/gen2convert/walker.py:
```diff
@@ -0,0 +1,329 @@
+        repoCfgPath = os.path.join(root, "repositoryCfg.yaml")
+        if os.path.exists(repoCfgPath):
+            with open(repoCfgPath, 'r') as f:
+                repoCfg = yaml.load(f)
```
Why is this not loaded with a `Config`?

TallJimbo replied:
I didn't want to assume that Gen3 config would remain compatible with Gen2 Config.

pschella commented on python/lsst/daf/butler/gen2convert/walker.py:
```diff
@@ -0,0 +1,329 @@
+                with open(mapperFilePath, 'r') as f:
+                    terms = f.read().strip().split(".")
+                    module = importlib.import_module(".".join(terms[:-1]))
+                    MapperClass = getattr(module, terms[-1])
```
Why not use `utils.doImport`?

TallJimbo replied:
I was just copying the code from Gen2; I'll try `utils.doImport` and see if it works.

pschella commented on python/lsst/daf/butler/gen2convert/walker.py:
```diff
@@ -0,0 +1,329 @@
+                parent = self.tryRoot(path)
+                if parent is None:
+                    raise ValueError("%s is a parent of %s, but is not a repository." % (parent, root))
+                repo.parents.append(parent)
```
Also add if ignored?

TallJimbo replied:
I think we can't really handle the case where a parent is ignored; I'll make that an explicit exception.

TallJimbo replied:
Actually, I think we have other checks for when the parent was necessary; I'll just remove ignored parents from the list here.

pschella commented on python/lsst/daf/butler/gen2convert/walker.py:
```diff
@@ -0,0 +1,329 @@
+                    raise ValueError("%s is a parent of %s, but is not a repository." % (parent, root))
+                repo.parents.append(parent)
+            # Make sure there are no cycles in the graph of parents.
+            repo._findCycles()
```
Doesn't this repeat the recursive cycle check for all it's parents? E.g. can't you use that `tryRoot` on it's parents has already not found any cycles? Probably not worth it, but still. Basically you want to know that the sets of parents and the current instance are disjoint recursively.

TallJimbo replied:
Because `tryRoot` short-circuits if `repo` is already in `self.found`, and we add the current `repo` to `self.found` before recursing, it won't find cycles.

pschella commented on python/lsst/daf/butler/gen2convert/walker.py:
```diff
@@ -0,0 +1,329 @@
+        log.info("%s: walking datasets.", repo.root)
+        for dirPath, dirNames, fileNames in os.walk(repo.root):
+            dirNames[:] = [dirName for dirName in dirNames
+                           if not self.tryRoot(os.path.join(dirPath, dirName))]
```
This looks quite expensive. If you are already in a repo, can a subdirectory still be a root?

TallJimbo replied:
Yes, though in practice this only happens under the `rerun` subdirectory of the repo root.  We can take advantage of that if we need to to speed things up in the future.

pschella commented on python/lsst/daf/butler/gen2convert/walker.py:
```diff
@@ -0,0 +1,329 @@
+
+        If self.MapperClass is None, set it from the MapperClass used by
+        parent repositories.  Raises ValueError if this is there is no single
+        MapperClass used by all parents.
```
Better to also have this in the standard docstring format with `Raises` section.

pschella commented on python/lsst/daf/butler/registries/sqlRegistry.py:
```diff
@@ -60,6 +60,33 @@ def __init__(self, config):
+        """
+        t = text(sql)
+        with self._engine.begin() as connection:
+            for row in connection.execute(t, **params):
```
This seems rather unsafe. What if the user doesn't read the comment properly and writes something other than a `SELECT` statement?

TallJimbo replied:
It certainly is, but I'd rather not deal with it now.  I'll add a TODO.

pschella commented on python/lsst/daf/butler/registries/sqlRegistry.py:
```diff
@@ -193,7 +221,8 @@ def addDataset(self, datasetType, dataId, run, producer=None):
             # A dataset is always associated with its Run collection
-            self.associate(run.collection, [datasetRef])
+            connection.execute(datasetCollectionTable.insert(),
+                               [{'dataset_id': datasetRef.id, 'collection': run.collection}])
```
:(

TallJimbo replied:
I've added a TODO for this, too.

pschella commented on tests/test_sqlDatabaseDict.py:
```diff
@@ -0,0 +1,108 @@
+            0: value(x=0, y="zero", z=0.0),
+            1: value(x=1, y="one", z=0.1),
+        }
+        self.checkDatabaseDict(value, data)
```
Let's name `value` -> `TestValue` since it is the type.

pschella commented on python/lsst/daf/butler/gen2convert/writer.py:
```diff
@@ -0,0 +1,316 @@
+        self.insertObservations(registry)
+        self.insertDatasetTypes(registry)
+        self.insertDatasets(registry, datastore)
+        # TODO: associate parent repo datasets with child repo Collections
```
Is this still open? Seems reasonably essential.

TallJimbo replied:
Yes, I'm afraid it is, and I remember it being hard but don't remember exactly why.  We can work around it with a custom call to `query` in the ci_hsc drivers for the conversion for now.

pschella commented on python/lsst/daf/butler/gen2convert/writer.py:
```diff
@@ -0,0 +1,316 @@
+            skyMapName = self.skyMapNames.get(sha1, None)
+            try:
+                existing, = registry.query("SELECT skymap FROM SkyMap WHERE sha1=:sha1",
+                                           sha1=sha1)
```
In general we probably don't want to rely on `query` too much, when other API calls should exist.

TallJimbo replied:
I completely agree.  I've been meaning to talk to you about how to add alternate unique lookup APIs for some DataUnits (SkyMap `sha1`, Patch `cell_x` and `cell_y`, Sensor `name`, probably some Camera-specific `Visit` or `Exposure` identifiers).

pschella commented on python/lsst/daf/butler/gen2convert/writer.py:
```diff
@@ -0,0 +1,316 @@
+                     "and does not already exist in the Registry.").format(sha1.hex())
+                )
+            log.info("Inserting SkyMap '%s' with sha1=%s", skyMapName, sha1.hex())
+            skyMap.register(skyMapName, registry)
```
Do we want the same skymap to have different names? Otherwise I don't really get why the call can't be `skyMap.register(registry)`.

pschella replied:
Or directly `addDataUnitEntry(...)`.

pschella commented on python/lsst/daf/butler/gen2convert/writer.py:
```diff
@@ -0,0 +1,316 @@
+                    "boresight_ra": visitInfo.getBoresightRaDec().getLongitude().asDegrees(),
+                    "boresight_dec": visitInfo.getBoresightRaDec().getLatitude().asDegrees(),
+                    "boresight_parallactic_angle": visitInfo.getBoresightParAngle().asDegrees(),
+                    "local_era": visitInfo.getLocalEra().asDegrees(),
```
Do we loose information with respect to `visitInfo` here?

TallJimbo replied:
Possibly, but I don't want to think too hard about this until we've gone through the schema review.

pschella commented on python/lsst/daf/butler/gen2convert/writer.py:
```diff
@@ -0,0 +1,316 @@
+        log = Log.getLogger("lsst.daf.butler.gen2convert")
+        for datasetType in self.datasetTypes.values():
+            # TODO: should put this "just make sure it exists" logic
+            # into registerDatasetType itself, and make it a bit more careful.
```
Definitely!

pschella reviewed (commented):
I'm assuming we can't really unittest any of this?

TallJimbo reviewed (commented):
I think I've addressed all review comments.  I'll squash and merge by Monday.

As for tests, I don't think it's a good use of time to try to cook up the kind of test data we'd need for small scale unit testing of this code in daf_butler.  Instead, I'm planning to add running the conversion to ci_hsc's SConstruct, and add checks in that package that we can access the datasets using the Gen3 butler and that this is the same as Gen2 access.  We need at least DM-14225 before that's worthwhile, though, and may want to add a few more concrete StorageClasses/Formatters first as well.
